### PR TITLE
[FLINK-8220] Implement set of network benchmarks

### PR DESCRIPTION
**Pull request by pnowojski**
This PR implements sets of network benchmarks that are intended to simulate various workloads for network stack. Benchmarks will be executed by `flink-benchmarks` (https://github.com/dataArtisans/flink-benchmarks/pull/3) project. We want to keep the code of the benchmarks here, since components they are using are not part of public api and we don't wont them braking unknowingly on changes in the flink code.

Example results (including performance fix #5104):
```
Benchmark                                   (channels)  (writers)   Mode  Cnt      Score      Error   Units
NetworkBenchmarkExecutor.networkThroughput           1          1  thrpt   10  23322.178 ±  261.931  ops/ms
NetworkBenchmarkExecutor.networkThroughput           1          4  thrpt   10  57279.697 ± 4152.714  ops/ms
NetworkBenchmarkExecutor.networkThroughput          50          1  thrpt   10  22927.629 ±  341.008  ops/ms
NetworkBenchmarkExecutor.networkThroughput          50          4  thrpt   10  60761.706 ± 1172.354  ops/ms
NetworkBenchmarkExecutor.networkThroughput        1000          1  thrpt   10  17765.148 ±  479.579  ops/ms
NetworkBenchmarkExecutor.networkThroughput        1000          4  thrpt   10  41030.963 ± 1149.041  ops/ms
```

## Verifying this change

This change adds two tests ensuring that benchmarks are compiling and executing as expected without an exceptions.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by zentol** [comment]
Why do you want to add these to flink when they are only used by an external repository? Wouldn't it make sense to consolidate all benchmarking code to the flink-benchmark repo?

**Comment by greghogan** [comment]
Or if `flink-benchmarks` does not need to be distributed then that code could be contributed to the main Flink repository.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/benchmark/SerializingLongReceiver.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    public SerializingLongReceiver(InputGate inputGate, int expectedRepetitionsOfExpectedRecord) {
+        super(expectedRepetitionsOfExpectedRecord);
+        this.reader = new MutableRecordReader<>(
+            inputGate,
+            new String[]{
+                EnvironmentInformation.getTemporaryFileDirectory()
+            });
+    }
+
+    protected void readRecords(long lastExpectedRecord) throws Exception {
```
Comment (casidiablo) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/benchmark/SerializingLongReceiver.java]: Add `@Override`?

**Comment by pnowojski** [comment]
@zentol putting this code in external repository would make those benchmarks very fragile, since they depend on Flink internals. Almost any refactor made in Flink's network stack would brake the external repository (of which most committers are not even aware of) and would require coordinated commits, which would be unmaintainable.

@greghogan hmmm, that's a very tempting option, however I'm not the one that must be convinced that such option is legally viable.

**Comment by pnowojski** [comment]
It seems like `commons-lang` doesn't have any issues with using `jmh` in tests: https://commons.apache.org/proper/commons-lang/dependencies.html

However @greghogan do you remember my previous attempt with JMH in Flink? https://github.com/apache/flink/pull/4323 Was it a false alarm?

**Comment by StephanEwen** [comment]
@greghogan That may or may not be possible, not 100% sure. Re-distributing it does certainly not work.

To my current understanding, the Apache guidelines say we would need to make sure that downstream consumers of Flink (also the source) don't accidentally pull in a GPL license. For weak copyleft, where consuming it only via linking, it is okay, but for strong copyleft, where accidentally linking it results in copyleft, I think the hurdles for that are pretty high. There are some gray zone areas (like MPL) where the understanding is that "optional" dependency is okay because it requires an explicit action by a downstream user (re-adding the dependency) and the potential damage is not too high.

I have seen a lot of cases first hand in the past months how many companies that want to use open source freak out as soon as they see an instance of GPL. We would make Flink's adoption easier if we did not have that.

My first suggestion would be to consolidate this in a separate `flink-benchmarks` repository. That would also keep the core Flink repository "clear" of GPL. I could check with Apache legal for a final answer.

**Comment by greghogan** [comment]
@pnowojski see FLINK-2848 and FLINK-2973. Also the BSD + Patents conversation along with Flink's dependence on Amazon's Kinesis library (likewise [Category X](https://www.apache.org/legal/resolved.html#category-x)). I now think we have been overly cautious in this regard, not necessarily a bad thing.

As @StephanEwen noted, the GPL-dependent code must be an [optional component](https://www.apache.org/legal/resolved.html#optional). We [do this](https://ci.apache.org/projects/flink/flink-docs-release-1.3/dev/connectors/kinesis.html) with Kinesis by not releasing convenience binaries but requiring the user to compile from source with an explicit flag. I think we are safe to include a `flink-benchmarks` module which is not depended on or included in the Flink distribution. The benchmarks are Apache licensed, only the dependence on `jmh` prohibits distribution, so can be included in the (source) release. These are performance, not integration, tests so would be run manually rather than during the build.

I agree with @StephanEwen's caution but we would not be releasing GPL code, it is not obvious that `jmh` is GPL licensed, and Java itself is GPL.

**Comment by StephanEwen** [comment]
@greg I think there is one more requirement by the ASF (not necessarily the Apache License itself), which is making sure that downstream consumers of Apache projects are always explicitly aware when they are consuming something (even when the build themselves) that has friction with the ASL 2.0.

My preferred solution would still be a separate repository. Executing benchmarks during tests is pretty tricky anyways, because their result depends so much on where the benchmark is executed, and what happens concurrently on the machine, is it a VM or container, etc.

**Comment by StephanEwen** [comment]
I also think being more on the cautious side is actually a good thing. Given how scared and confused many people and companies are with respect to open source and accidentally infringement of IP, being cautious is good, in my opinion.

There is an interesting comment from Ted Dunning on a related question, basically saying that if things could be a problem or not, think if some cautious/picky users might perceive them as a problem:
https://lists.apache.org/thread.html/fc3992c13cc30f889c820d1cfd6be61b63a5be4efa7c9101262474c9@%3Clegal-discuss.apache.org%3E

**Comment by greghogan** [comment]
@StephanEwen thanks for the link and thoughtful discussion. Including `jmh` benchmarks has come up multiple times. Do you see any path to including such a benchmark module in the main Flink repo?

**Comment by StephanEwen** [comment]
I having a `flink-benchmarks` repo is not a solution that the community likes, we could check whether adding it in `test` scope is okay, if we ensure we do not publish a "test-jar" - that way there is no redistribution and no user could accidentally pull it in.

**Comment by zentol** [comment]
Skipping the `maven-deploy-plugin` should accomplish that:

```
	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-deploy-plugin</artifactId>
				<configuration>
					<skip>true</skip>
				</configuration>
			</plugin>
		</plugins>
	</build>
```

**Comment by StephanEwen** [comment]
Would be great if someone could check with Apache legal if this is actually in line with policies. Either on the `LEGAL` Jira, or on the legal-discuss@apache.org mailing list.

**Comment by NicoK** [comment]
merged in c816191113d813156467f3e33856636ef0bcce38
@pnowojski can you close this PR?

**Pull request closed without merging**